applications/ShapeOptimizationApplication/custom_utilities/response_functions/water_drain_response_function_utility.cpp

```diff
@@ -0,0 +1,454 @@
+// ==============================================================================
+//  KratosShapeOptimizationApplication
+//
+//  License:         BSD License
+//                   license: ShapeOptimizationApplication/license.txt
+//
+//  Main authors:    Schmölz David, https://github.com/dschmoelz
+//
+// ==============================================================================
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "custom_utilities/response_functions/water_drain_response_function_utility.h"
+#include "custom_utilities/geometry_utilities.h"
+#include "shape_optimization_application.h"
+#include "utilities/variable_utils.h"
+#include "utilities/parallel_utilities.h"
+#include "utilities/reduction_utilities.h"
+#include "includes/global_pointer_variables.h"
+#include "includes/model_part.h"
+#include "processes/find_global_nodal_neighbours_for_entities_process.h"
+#include "processes/find_global_nodal_entity_neighbours_process.h"
+#include "utilities/builtin_timer.h"
+#include "utilities/openmp_utils.h"
+
+
+// ==============================================================================
+
+namespace Kratos
+{
+
+WaterDrainResponseFunctionUtility::WaterDrainResponseFunctionUtility(ModelPart& rModelPart, Parameters ResponseSettings)
+	: mrModelPart(rModelPart), mGeometryUtilities(GeometryUtilities(mrModelPart))
+{
+	const std::size_t domain_size = mrModelPart.GetProcessInfo()[DOMAIN_SIZE];
+	KRATOS_ERROR_IF(domain_size != 3) << "WaterDrainResponseFunctionUtility can only be used on 3D geometries!" << std::endl;
+
+	mGravityDirection = ResponseSettings["gravity_direction"].GetVector();
+	const double direction_norm = norm_2(mGravityDirection);
+	KRATOS_ERROR_IF(direction_norm < std::numeric_limits<double>::epsilon()) << "WaterDrainResponseFunctionUtility: 'main_direction' vector norm is 0!" << std::endl;
```

Comment:
std::numeric_limits<double>::epsilon()  might be too small, consider use 1e-16 or so ...

Reply:
changed to 1e-16